src/frontend/screens/MapScreen/GPSPermissions/GPSPermissionsModal.tsx

```diff
@@ -1,34 +1,84 @@
-import React, {useEffect, useState} from 'react';
-import {GPSPermissionsDisabled} from './GPSPermissionsDisabled';
+import React from 'react';
+import {GPSForegroundPermissionDisabled} from './GPSForegroundPermissionDisabled';
 import {GPSPermissionsEnabled} from './GPSPermissionsEnabled';
 import * as Location from 'expo-location';
 import {useGPSModalContext} from '../../../contexts/GPSModalContext';
 import {useTabNavigationStore} from '../../../hooks/useTabNavigationStore';
 import {BottomSheetModal, BottomSheetView} from '@gorhom/bottom-sheet';
 import {TAB_BAR_HEIGHT} from '../../../Navigation/Stack/AppScreens';
-import {StyleSheet} from 'react-native';
-import {useFocusEffect} from '@react-navigation/native';
+import {StyleSheet, Linking, View} from 'react-native';
+import {GPSBackgroundPermissionDisabled} from './GPSBackgroundPermissionDisabled';
+import {Loading} from '../../../sharedComponents/Loading';
+
+const handleOpenSettings = () => {
+  Linking.sendIntent('android.settings.LOCATION_SOURCE_SETTINGS');
+};
 
 export const GPSPermissionsModal = React.memo(() => {
   const {setCurrentTab} = useTabNavigationStore();
-  const [backgroundStatus] = Location.useBackgroundPermissions();
-  const [foregroundStatus] = Location.useForegroundPermissions();
 
-  const [isGranted, setIsGranted] = useState<boolean | null>(null);
+  const [foregroundPermission, setForegroundPermission] =
+    React.useState<Location.LocationPermissionResponse | null>(null);
+  const [backgroundPermission, setBackgroundPermission] =
+    React.useState<Location.LocationPermissionResponse | null>(null);
   const {bottomSheetRef} = useGPSModalContext();
 
-  useEffect(() => {
-    if (backgroundStatus && foregroundStatus && isGranted === null) {
-      setIsGranted(backgroundStatus.granted && foregroundStatus.granted);
-    }
-  }, [backgroundStatus, foregroundStatus, isGranted]);
+  React.useEffect(() => {
+    Location.getForegroundPermissionsAsync().then(permission =>
+      setForegroundPermission(permission),
+    );
+
+    Location.getBackgroundPermissionsAsync().then(permission =>
+      setBackgroundPermission(permission),
+    );
+  }, []);
 
   const onBottomSheetDismiss = () => {
     setCurrentTab('Map');
   };
-  useFocusEffect(() => {
-    return () => bottomSheetRef?.current?.close();
-  });
+
+  async function askForegroundLocationPermission() {
+    if (foregroundPermission!.canAskAgain) {
+      const permission = await Location.requestForegroundPermissionsAsync();
+      setForegroundPermission(permission);
+    } else {
+      handleOpenSettings();
+    }
+  }
+
+  async function askBackgroundLocationPermission() {
+    if (backgroundPermission!.canAskAgain) {
+      const permission = await Location.requestBackgroundPermissionsAsync();
+      setBackgroundPermission(permission);
+    } else {
+      handleOpenSettings();
+    }
+  }
```

Comment:
nit: would inline these to where they're used so you can avoid the non-null operator usage e.g. `backgroundPermission!.canAskAgain` becomes `backgroundPermission.canAskAgain`